rust-runtime/aws-smithy-http-server-python/src/server.rs

```diff
@@ -424,19 +424,30 @@ event_loop.add_signal_handler(signal.SIGINT,
     /// it starts the Lambda handler on the current process.
     fn run_lambda_handler(&mut self, py: Python) -> PyResult<()> {
         let event_loop = self.configure_python_event_loop(py)?;
+        // Register signals on the Python event loop.
+        self.register_python_signals(py, event_loop.to_object(py))?;
+
         let service = self.build_and_configure_service(py, event_loop)?;
-        let rt = runtime::Builder::new_multi_thread()
-            .enable_all()
-            .build()
-            .expect("unable to start a new tokio runtime for this process");
-        rt.block_on(async move {
-            let handler = LambdaHandler::new(service);
-            let lambda = lambda_http::run(handler);
-            tracing::debug!("starting lambda handler");
-            if let Err(err) = lambda.await {
-                tracing::error!(error = %err, "unable to start lambda handler");
-            }
+
+        // Spawn a new background [std::thread] to run the application.
+        tracing::trace!("start the tokio runtime in a background task");
+        thread::spawn(move || {
+            let rt = runtime::Builder::new_multi_thread()
+                .enable_all()
+                .build()
+                .expect("unable to start a new tokio runtime for this process");
+            rt.block_on(async move {
+                let handler = LambdaHandler::new(service);
+                let lambda = lambda_http::run(handler);
+                tracing::debug!("starting lambda handler");
+                if let Err(err) = lambda.await {
+                    tracing::error!(error = %err, "unable to start lambda handler");
+                }
+            });
```

Comment:
What is the reason for moving from blocking directly on the Lambda await to using a thread and blocking on the Python event loop? Is it to handle signal properly?

Reply:
`asyncio` runtime doesn't work property if it does not control the main thread: https://docs.rs/pyo3-asyncio/latest/pyo3_asyncio/#pythons-event-loop-and-the-main-thread. Basically without this it runs into deadlock or something and hangs forever if we `await` on a Python coroutine. I borrowed this logic from `start_hyper_worker` btw, I didn't understand at first why that method is designed that way but now I understand 😄

Reply:
Got it, makes sense.

Reply:
This confused me too until I opened this conversation. Should we add a brief explanation/the link above as a comment?